src/miniscript/iter.rs

```diff
@@ -0,0 +1,82 @@
+// Miniscript
+// Written in 2020 by
+//     Dr Maxim Orlovsky <orlovsky@pandoracore.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+use std::{sync::Arc, ops::Deref};
+use super::{Miniscript, MiniscriptKey};
+
+impl<Pk: MiniscriptKey> Miniscript<Pk> {
+    pub fn iter(&self) -> MiniscriptIter<Pk> {
+        MiniscriptIter::new(self)
+    }
+
+    pub fn branches(&self) -> Vec<&Miniscript<Pk>> {
+        use Terminal::*;
+        match &self.node {
+            Pk(_) |
+            PkH(_) |
+            ThreshM(_, _) =>
+                vec![],
+
+            Alt(node) |
+            Swap(node) |
+            Check(node) |
+            DupIf(node) |
+            Verify(node) |
+            NonZero(node) |
+            ZeroNotEqual(node) =>
+                vec![node],
+
+            AndV(node1, node2) |
+            AndB(node1, node2) |
+            OrB(node1, node2) |
+            OrD(node1, node2) |
+            OrC(node1, node2) |
+            OrI(node1, node2) =>
+                vec![node1, node2],
```

Comment:
This is where the linear space comes from.

Reply:
There is no need to create a vector here; just have the stack remember the last-used child index.